### ISSUE TITLE
@User-Defined Variables ignored 

### ISSUE THREAD

**Issue by kael** [mysqljs/mysql#1751]
I'm trying to run the [How To Get Top N Rows per Each Group](http://www.sqlines.com/mysql/how-to/get_top_n_each_group) query which makes use of @user-defined variables.

Running the query with `mysql` and `node-mysql` doesn't return the same result, the @user-defined variables seems ignored.

- `mysql`:
```sql
SELECT city, country, population
FROM
    (SELECT
       city,
       country,
       population, 
       @country_rank := IF(@current_country = country, @country_rank + 1, 1) AS country_rank,
       @current_country := country
     FROM cities
     ORDER BY country, population DESC
    ) ranked
WHERE country_rank <= 2;
```
```sql
+-------------+----------------+------------+
| city        | country        | population |
+-------------+----------------+------------+
| Paris       | France         |    2181000 |
| Marseille   | France         |     808000 |
| London      | United Kingdom |    7825300 |
| Birmingham  | United Kingdom |    1016800 |
| New York    | United States  |    8175133 |
| Los Angeles | United States  |    3792621 |
+-------------+----------------+------------+
6 rows in set (0,00 sec)
```
- `node-mysql`:
```js
var sqlServer = {
    host: 'localhost',
    user: 'user',
    password: 'pass',
    database: 'test',
    port: 3306,
    charset: 'utf8',
    multipleStatements: true
};
var query = `
   SELECT city, country, population
   FROM
     (SELECT city, country, population, 
                  @country_rank := IF(@current_country = country, @country_rank + 1, 1) AS country_rank,
                  @current_country := country 
       FROM cities
       ORDER BY country, population DESC
     ) ranked
   WHERE country_rank <= 2;
`;
sql.connect({server: sqlServer}).then(() =>
  sql.query(query).then(result => console.log(result))
);
```

```js
[ RowDataPacket { city: 'Paris', country: 'France', population: 2181000 },
  RowDataPacket { city: 'Marseille', country: 'France', population: 808000 },
  RowDataPacket { city: 'Lyon', country: 'France', population: 422000 },
  RowDataPacket {
    city: 'London',
    country: 'United Kingdom',
    population: 7825300 },
  RowDataPacket {
    city: 'Birmingham',
    country: 'United Kingdom',
    population: 1016800 },
  RowDataPacket { city: 'Leeds', country: 'United Kingdom', population: 770800 },
  RowDataPacket {
    city: 'New York',
    country: 'United States',
    population: 8175133 },
  RowDataPacket {
    city: 'Los Angeles',
    country: 'United States',
    population: 3792621 },
  RowDataPacket {
    city: 'Chicago',
    country: 'United States',
    population: 2695598 } ]

```

Am I missing something ?

`node-mysql`: _^2.13.0_
`node`: _v6.10.3_
`npm`: _3.10.10_
`os`: _Ubuntu 16.04.2 LTS 4.4.0-78-generic_

**Comment by sidorares** [comment]
@kael  how do you get `sql` object? It does not look like mysqljs/mysql connection - `Connection.connect` returns undefined and in your case it looks like a promise

**Comment by dougwilson** [comment]
That's because `node-mysql` is a different module: https://www.npmjs.com/package/node-mysql @kael looks like you filed your issue in the wrong place.

This module is `mysql`, which it looks like you wait was giving you the right result, and `node-mysql` wasn't.

**Comment by kael** [comment]
@sidorares The `sql` object is an instance of a class I wrote.
@dougwilson Actually, I'm not using `node-mysql` but `"mysql": "^2.13.0"` (from `package.json`) indeed - didn't know they were different, confusing though.

Could you please give a try with the above query on your side ?

**Comment by sidorares** [comment]
@kael can you put together everything (including schema creation and data population) into one single script ( with mysqljs/mysql as only dependency ) and I'll try to debug

**Comment by dougwilson** [comment]
Sorry, I guess I'm confused by you (1) saying node-mysql which is a different module and (2) as pointed out, your code isn't even something that would be working with this module.

So yea, we can look if we had the schema and data to run against and your full script.

**Comment by kael** [comment]
Here's a script with table creation, data population and the query:

```sql

var mysql = require('mysql');
var connection = mysql.createConnection({
  host: 'localhost',
  user: 'user',
  password: 'pass',
  database: 'test',
  port: 3306,
  charset: 'utf8',
  multipleStatements: true
});

connection.connect();

var dropTable = `DROP TABLE IF EXISTS cities;`,
    createTable = `
      CREATE TABLE cities
        (
          city VARCHAR(80),
          country VARCHAR(80),
          population INT
        );
    `,
    insertTable = `
       INSERT INTO cities VALUES ('New York', 'United States', 8175133); 
       INSERT INTO cities VALUES ('Los Angeles', 'United States', 3792621); 
       INSERT INTO cities VALUES ('Chicago', 'United States', 2695598); 
     
       INSERT INTO cities VALUES ('Paris', 'France', 2181000);
       INSERT INTO cities VALUES ('Marseille', 'France', 808000);
       INSERT INTO cities VALUES ('Lyon', 'France', 422000);
     
       INSERT INTO cities VALUES ('London', 'United Kingdom', 7825300);
       INSERT INTO cities VALUES ('Birmingham', 'United Kingdom', 1016800);
       INSERT INTO cities VALUES ('Leeds', 'United Kingdom', 770800);
    `;

connection.query(dropTable, (error, results, fields) => {
  // console.log('dropTable', results);
});
connection.query(createTable, (error, results, fields) => {
  // console.log('createTable', results);
});
connection.query(insertTable, (error, results, fields) => {
  // console.log('insertTable', results);
});

var selectTopNRows = `
  SELECT city, country, population
  FROM
      (SELECT
         city,
         country,
         population, 
         @country_rank := IF(@current_country = country, @country_rank + 1, 1) AS country_rank,
         @current_country := country
       FROM cities
       ORDER BY country, population DESC
      ) ranked
  WHERE country_rank <= 2;
`;

connection.query(selectTopNRows, (error, results, fields) => {
  console.log('selectTopNRows', results);
});

connection.end();
```

The result can be compared with the one from the example [page](http://www.sqlines.com/mysql/how-to/get_top_n_each_group).

**Comment by sidorares** [comment]
I have exactly the same results using mysqljs/mysql and MySQL workbench @kael  - please double check the data and expected result. Looks like everything works as expected

**Comment by kael** [comment]
Erm , I don't get that. :confused:

With _MySQL Workbench_ I get a 6 rows result, same as with _mysql_ and from the example:

![mysql-workbench](https://i.imgur.com/IWDW9AD.png)
![mysql](https://i.imgur.com/peWuj9H.png)

With _mysqljs_, I get a 9 rows result (for that query):

```js
var selectTopNRows = `
  SELECT city, country, population
  FROM
      (SELECT
         city,
         country,
         population, 
         @country_rank := IF(@current_country = country, @country_rank + 1, 1) AS country_rank,
         @current_country := country
       FROM cities
       ORDER BY country, population DESC
      ) ranked
  WHERE country_rank <= 2;
`;
console.log('=========================');
console.log(selectTopNRows);
console.log('=========================');
connection.query(selectTopNRows, (error, results, fields) => {
  console.log('selectTopNRows', results);
});
```

![mysql-workbench](https://i.imgur.com/zIBEDcy.png)

It's the same query copied-pasted in all cases. I've double checked it's using the correct table from the correct database.

**Comment by sidorares** [comment]
I get 9 with workbench

**Comment by kael** [comment]
@sidorares But a 9 rows result is incorrect compared to the [expected one](http://www.sqlines.com/mysql/how-to/get_top_n_each_group#mysql-query-to-get-top-2).

I'm using _MySQL Workbench 6.0.8+dfsg-2_ on _Ubuntu Trusty 14.04.5 LTS_.

**Comment by dougwilson** [comment]
I also got 9 results when I tried it like @sidorares . @kael what version of your MySQL server are you running? Are you connecting to the same server for both the workbench trial and this module trial? This module doesn't perform any logic, just sends the exact string to your MySQL server to run and those row packets are just whatever your MySQL server sent back, so we need to narrow down what is different between your workbench connection and the connection through this module.

**Comment by dougwilson** [comment]
FWIW, I found a SO post (https://stackoverflow.com/questions/29425805/mysql-variable-and-if-statement) that doesn't involve Node.js in any way, but people had this same behavior: some were seeing 6 results, some 9 results for the query on that sqllines.com site.

**Comment by kael** [comment]
This is the version of the MySQL server running remotely : `mysql  Ver 14.14 Distrib 5.7.18, for Linux (x86_64) using  EditLine wrapper`. I'm acessing it over SSH with _Workbench_.

But I've actually spotted something :tada: : the query fails the first time but succeeds the second and following times, with both _mysqljs_ and _Workbench_.

The following code with the query ran multiple times :
```sql
var selectTopNRows = `
  SELECT city, country, population
  FROM
      (SELECT
         city,
         country,
         population, 
         @country_rank := IF(@current_country = country, @country_rank + 1, 1) AS country_rank,
         @current_country := country
       FROM cities
       ORDER BY country, population DESC
      ) ranked
  WHERE country_rank <= 2;
`;
connection.query(selectTopNRows, (error, results, fields) => {
  console.log('selectTopNRows 1\n', results, '\n\n', "=> results length", results.length, '\n');
});
connection.query(selectTopNRows, (error, results, fields) => {
  console.log('selectTopNRows 2\n', results, '\n\n', "=> results length", results.length, '\n');
});
connection.query(selectTopNRows, (error, results, fields) => {
  console.log('selectTopNRows 3\n', results, '\n\n', "=> results length", results.length, '\n');
});
connection.end();
```

Outputs:
```sql
selectTopNRows 1
 [ RowDataPacket { city: 'Paris', country: 'France', population: 2181000 },
  RowDataPacket { city: 'Marseille', country: 'France', population: 808000 },
  RowDataPacket { city: 'Lyon', country: 'France', population: 422000 },
  RowDataPacket {
    city: 'London',
    country: 'United Kingdom',
    population: 7825300 },
  RowDataPacket {
    city: 'Birmingham',
    country: 'United Kingdom',
    population: 1016800 },
  RowDataPacket { city: 'Leeds', country: 'United Kingdom', population: 770800 },
  RowDataPacket {
    city: 'New York',
    country: 'United States',
    population: 8175133 },
  RowDataPacket {
    city: 'Los Angeles',
    country: 'United States',
    population: 3792621 },
  RowDataPacket {
    city: 'Chicago',
    country: 'United States',
    population: 2695598 } ] 

 => results length 9 

selectTopNRows 2
 [ RowDataPacket { city: 'Paris', country: 'France', population: 2181000 },
  RowDataPacket { city: 'Marseille', country: 'France', population: 808000 },
  RowDataPacket {
    city: 'London',
    country: 'United Kingdom',
    population: 7825300 },
  RowDataPacket {
    city: 'Birmingham',
    country: 'United Kingdom',
    population: 1016800 },
  RowDataPacket {
    city: 'New York',
    country: 'United States',
    population: 8175133 },
  RowDataPacket {
    city: 'Los Angeles',
    country: 'United States',
    population: 3792621 } ] 

 => results length 6 

selectTopNRows 3
 [ RowDataPacket { city: 'Paris', country: 'France', population: 2181000 },
  RowDataPacket { city: 'Marseille', country: 'France', population: 808000 },
  RowDataPacket {
    city: 'London',
    country: 'United Kingdom',
    population: 7825300 },
  RowDataPacket {
    city: 'Birmingham',
    country: 'United Kingdom',
    population: 1016800 },
  RowDataPacket {
    city: 'New York',
    country: 'United States',
    population: 8175133 },
  RowDataPacket {
    city: 'Los Angeles',
    country: 'United States',
    population: 3792621 } ] 

 => results length 6 

```

The first `selectTopNRows 1` rows set is wrong, but the following are correct.

Similarly with _Workbench_, if you run the query multiple times, the first output will be of 9 rows, the following ones of 6 rows.

I'm going to test on my side how it can be circumvented in day-to-day use.

**Comment by kael** [comment]
The same bug happens with the mysql shell, though.

**Comment by dougwilson** [comment]
So we're able to reproduce the behavior through workbench and the command line, both of which are not this module, so it's safe to say the issue has nothing to do with this module. In fact, the SO post got me thinking, since people kept saying that you need to first declare the two variables. Reading the MySQL manual, this is in fact necessary. You can see this in the section on user variables (https://dev.mysql.com/doc/refman/5.7/en/user-variables.html). Some select quotes:

> If you refer to a variable that has not been initialized, it has a value of NULL and a type of string. 

> Another issue with assigning a value to a variable and reading the value within the same non-SET statement is that the default result type of a variable is based on its type at the start of the statement.

Basically the reason why the first run of the query is wrong is because your variable `@country_rank` is not initialized yet. This means that the type of the variable for the entire statement run is a string, not an integer, so the add operations do not work at all in the first run, which is why all rows are returned, since the `<= 2` is true for all the rows.

Once you ran it once (or, alternatively as in the SO post, you declare the two variables with the correct types), then your query will return the expected results.

I hope this helps.

**Comment by kael** [comment]
Thanks a lot for your help. Running the query once just after socket connection helps to circumvent the problem. Cheers.
